origins_qa/src/Controller/QaApiController.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\origins_qa\Controller;
+
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\Core\File\FileSystemInterface;
+use Drupal\Core\Logger\LoggerChannelFactory;
+use Drupal\Core\Site\Settings;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Provides endpoints for the QA API service.
+ */
+final class QaApiController extends ControllerBase {
+
+  /**
+   * The filepath to the invalid token list.
+   *
+   * @var string
+   */
+  protected $invalidTokensFilepath;
+
+  /**
+   * The current request.
+   *
+   * @var \Symfony\Component\HttpFoundation\Request
+   */
+  protected $request;
+
+  /**
+   * The logger channel.
+   *
+   * @var \Drupal\Core\Logger\LoggerChannel
+   */
+  protected $logger;
+
+  /**
+   * Constructs a QaEndpointController object.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The current request.
+   *
+   * @param \Drupal\Core\Logger\LoggerChannelFactory $logger
+   *   The logger channel factory service.
+   */
+  public function __construct(Request $request, LoggerChannelFactory $logger) {
+    $this->request = $request;
+    $this->logger = $logger->get('origins_qa');
+    $this->invalidTokensFilepath = Settings::get('file_private_path') . '/origins_qa_invalid_tokens.txt';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('request_stack')->getCurrentRequest(),
+      $container->get('logger.factory'),
+    );
+  }
+
+  /**
+   * Enable QA accounts.
+   */
+  public function setQaUsersStatus($status, $token) {
+    // Check if the token is in the invalid list.
+    if (file_exists($this->invalidTokensFilepath)) {
+      $invalid_tokens = str_getcsv(file_get_contents($this->invalidTokensFilepath));
```

Comment:
What happens to the invalid tokens file? Is it junked out periodically? If I spam the route with bogus requests could I fill up the file and potentially the mount it sits on?

Reply:
Nope and yup I'd thought it would be easy to spam that and eat up the disk space so I'm looking at flood control for routes.